language-design.md

```diff
@@ -17,7 +17,78 @@ For each quote, describe what it was about the quote that led you pick it.
 
 **Response**
 
-
+> I have said in the past, and will say now, that I think it would be a good thing
+> for the Java programming language to ... let the user define overloaded operators.
+> Just as a user can code methods that can be used in just the same way as
+> methods that are built in, the user ought to have a way to define operators for user
+> defined classes that can be used in just the same way as operators that are built in. [Steele, 1998]
+
+I never really considered the importance of operator overloading before, but Steele makes a very convincing argument for them:
+opertor overloading allows an extension of a language (e.g., a class or package) to be as concise, familiar, and 'primitive'
+as the underlying language. For example, in languages such as Java where operator overloading is not supported,
+users must write an `add` function to represent the addition of two objects; the resulting code looks something like
+`add(obj1, obj2)`, which gets the job done but isn't as simple and pleasing as the operator overloaded alternative `obj1 + obj2`.
```

Comment:
Well, operator overloading is just syntactic sugar and doesn't actually solve this issue. Subclassing also wouldn't solve it. It's actually a problem with the fact that the operator is only associated with one of the two objects in question.

In C++, A == B doesn't mean A.equals(B), but rather equals(A, B) [as a best practice, anyways]. Then, overloading solves the issue because the new equality operator will be more specific. The reason Java can't do this is because it doesn't have external functions in that sense.

Python actually would have the same issue as Java, despite having operator overloading, but fixes it via MAGIC, namely the NotImplemented constant. If A.__eq__(B) returns NotImplemented, then it tries B.__eq__(A). If that also returns NotImplemented, then it falls back on a default which is true equivalence (physical memory location). This means that binary operators are treated differently than other functions, and also makes implementing operators much more difficult if you want full generality.

So I think it's kind of a fundamental problem with making operators be methods at all. Which goes to show how much a seemingly straightforward conceptual choice can lead to unforeseen issues.